packages/flame/lib/src/experimental/layout_component.dart

```diff
@@ -0,0 +1,226 @@
+import 'package:collection/collection.dart';
+import 'package:flame/components.dart';
+import 'package:flame/image_composition.dart';
+import 'package:flutter/rendering.dart';
+
+enum Direction { horizontal, vertical }
+
+/// Superclass for linear layouts.
+/// A re-layout is performed when
+///  - a change in this component's children takes place
+///  - the [gap] parameter is changed
+///  - the [size] parameter is changed
+///  - the [mainAxisAlignment] parameter is changed
+///  - the [crossAxisAlignment] parameter is changed
+///
+/// [gap] is ignored when the [mainAxisAlignment] is set to one of:
+///  - [MainAxisAlignment.spaceAround]
+///  - [MainAxisAlignment.spaceBetween]
+///  - [MainAxisAlignment.spaceEvenly]
+///
+/// Notes:
+///  - currently, [CrossAxisAlignment.baseline] is unsupported, and behaves
+/// exactly like [CrossAxisAlignment.start].
+abstract class LayoutComponent extends PositionComponent {
+  LayoutComponent({
+    required this.direction,
+    required CrossAxisAlignment crossAxisAlignment,
+    required MainAxisAlignment mainAxisAlignment,
+    required super.position,
+    required super.size,
+    required double gap,
+    required super.children,
+  })  : _crossAxisAlignment = crossAxisAlignment,
+        _mainAxisAlignment = mainAxisAlignment,
+        _gap = gap;
+  final Direction direction;
+  CrossAxisAlignment _crossAxisAlignment;
+
+  CrossAxisAlignment get crossAxisAlignment => _crossAxisAlignment;
+
+  set crossAxisAlignment(CrossAxisAlignment value) {
+    _crossAxisAlignment = value;
+    layoutChildren();
+  }
+
+  MainAxisAlignment _mainAxisAlignment;
+
+  MainAxisAlignment get mainAxisAlignment => _mainAxisAlignment;
+
+  set mainAxisAlignment(MainAxisAlignment value) {
+    _mainAxisAlignment = value;
+    layoutChildren();
+  }
+
+  double _gap;
+
+  /// gap between components
+  double get gap => _gap;
+  set gap(double gap) {
+    _gap = gap;
+    layoutChildren();
+  }
+
+  @override
+  set size(Vector2 size) {
+    super.size = size;
```

Comment:
Listen to the size instead, since there are many ways to change its values (like `size.x = ...`)